### DISCUSSION TITLE
CI Improvements

### DISCUSSION THREAD

**Discussion by keyseven123** [nebulastream/nebulastream#162] [Ideas]
With #23 and #22 , we already have taken a step into improving the CI / the build process. This discussion acts as a brainstorming and idea collector of what we can do next to further enhance our CI. Once we have decided on a particular issue/point, the point/issue will be added to the following list:

- TBA

**Comment by keyseven123** [comment]
# Format-Checks

Currently, a CI job triggers a docker container that runs a bash script, calling a `cmake` target that calls a python script. 
One idea is to provide a bash script that gets called directly in the github workflow. This script can then be also called in a pre-commit hook

**Reply by fwc** [reply]
https://github.com/nebulastream/nebulastream/pull/5025 contains a simple-ish parallel script :)

**Reply by alepping** [reply]
Also, the format-check currently takes longer than our tests (~2min30s) it would be great to improve this.
![Screenshot from 2024-08-01 08-51-07](https://github.com/user-attachments/assets/2c683d14-2ec1-415b-9d57-f8c4786fc148)

**Reply by fwc** [reply]
Afaics thisi is mostly due to configure time overhead; especially the download of the pre-build dependencies. After the dependencies refactor, this might be much faster.

> a CI job triggers a docker container that runs a bash script, calling a cmake target that calls a python script.

But I think reducing this indirection is generally very nice.

**Comment by adrianmichalke** [comment]
# Merge Queue
A nice to have would be to have a merge queue. Seems like this would take from us the burden of organizing us in the `merge-my-pr` channel.
Not sure how much much work this is though.

[Github Docs](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue)
[Clickhouse Impl](https://github.com/ClickHouse/ClickHouse/blob/master/.github/workflows/merge_queue.yml)

**Edit:** nvm, seems like we need to be open source for github to support it.

**Reply by alepping** [reply]
> Edit: nvm, seems like we need to be open source for github to support it.

I would imagine that there are ways to automate this manually though using the gh CLI.
Generated sketch:
```python
import subprocess
import json
import time

def get_pr_status(pr_number):
    result = subprocess.run(['gh', 'pr', 'view', str(pr_number), '--json', 'mergeable,mergeStateStatus'], capture_output=True, text=True)
    return json.loads(result.stdout)

def merge_pr(pr_number):
    subprocess.run(['gh', 'pr', 'merge', str(pr_number), '--merge'])

def main():
    # This would be your queue of PR numbers
    pr_queue = [123, 124, 125]  # Replace with actual PR numbers
    
    while pr_queue:
        for pr_number in pr_queue:
            status = get_pr_status(pr_number)
            
            if status['mergeable'] and status['mergeStateStatus'] == 'CLEAN':
                print(f"Merging PR #{pr_number}")
                merge_pr(pr_number)
                pr_queue.remove(pr_number)
                break
            else:
                print(f"PR #{pr_number} is not ready to merge. Moving to next PR.")
        
        time.sleep(60)  # Wait for 1 minute before checking again

if __name__ == "__main__":
    main()
```

**Comment by adrianmichalke** [comment]
# Documentation docs.nebula.stream
I would like to have the documentation website in our repo. Then we would need a CI job that publishes a new version whenever the `documentation` or `docs` folder has changed.

**Comment by alepping** [comment]
# Pragma Once
Clickhouse and DuckDB use `#pragma once` showing that it works well in praxis, even for larger projects with strict compatibility constraints. `#pragma once` is not part of the standard and was not well supported by compilers more than a decade ago (see https://stackoverflow.com/questions/1143936/pragma-once-vs-include-guards). However, even in 2009 (look at the top comment in the link), there was already great  support and several bugs were fixed over time.
An argument against `#pragma once` from the above link is that the files with the same basename might both be suppressed when using `#pragma once`. However, [here](https://www.reddit.com/r/cpp/comments/4cjjwe/comment/d1j04te/) people make good arguments against it.
Nowadays, there really seems to be no very convincing argument against `#pragma once`.

**Comment by alepping** [comment]
# Pre-Commit Hook
Currently, we need to manually execute fix-guard, format, and the comment check A pre-commit hook can automate this for us.
A few considerations:
+ (+) automated and streamlined application of our formatting
+ (+) potentially reduces CI overhead because a developer pushed to an issue that has an open PR without applying formatting
- (-) formatting affects commits (this is only true for `format` and since the codebase now follows the new format, the impact on individual commits should be negligible.)

Here is a draft for a pre-commit hook:
```bash
#!/bin/bash

# Pre-push hook for CMake project
# This script runs formatting, guard fixing, and comment checking before pushing

# Store the name of the current branch
current_branch=$(git rev-parse --abbrev-ref HEAD)

# Function to run a CMake target
run_cmake_target() {
    target=$1
    echo "Running $target..."
    if ! cmake --build cmake-build-debug --target "$target" -j 10; then
        echo "Error: $target failed. Push aborted."
        exit 1
    fi
}

# Run the CMake targets
run_cmake_target "check-comment-format"
run_cmake_target "format"
run_cmake_target "fix-guards"

# If all checks pass, allow the push
echo "All checks passed. Proceed with push."
exit 0

```
The script assumes that `format` and `fix-guards` do not fail.
The comment-format check is executed first. If it fails, then the commit aborts. Staged files still remain staged after aborting the commit.

**Reply by adrianmichalke** [reply]
Cool! I guess everyone has to include it by themselves into their .git/hooks/pre-commit

Would that be a file in our repo, a section in our code guide docs or just a suggestion here?

**Reply by fwc** [reply]
Afaics, this hook does not handle cases like when some files are staged and some files are not. I think [pre-commit](https://pre-commit.com/) has most of this hashed out, so this might be a more robust way...

(And pre-commit would also have this rather "standard" `.pre-commit-config.yaml` in the repo, so that users only need to install pre-commit and then run `pre-commit install` to register with git.

**Comment by ls-1801** [comment]
## Run format checks on every commit Individually

This ensures that each commit can be dropped without breaking formatting invariants, it also prevents commits like `fix format` appended to every pr.

**Comment by keyseven123** [comment]
# Running tests with all cmake options enabled and disabled
We should make sure that NebulaStream can be run with all cmake options enabled and disabled. This way, we can ensure that a simple 
```
git clone
mkdir build && cd build && cmake .. 
make test_default
```
works all the time. 
We might only run this only in our nightly CI job

**Comment by adrianmichalke** [comment]
# CI Checks per PR
Some additional PR checks that we want to integrate:
 **1.** There are no TODOs without issue number --> always: `/// TODO(#XX): yyy`
 **2.** All TODOs referring to the issue(s) to be closed are removed --> master after merging PR for issue N does not contain any TODO(#N) 
 **3.** The commits follows our structure --> follow conventional commit and contain issue nr

**Comment by keyseven123** [comment]
# Checking for only allowing relative paths

As part of our coding guidelines, we agreed on using relative paths for includes only. We should check this similar to our format and license check

**Comment by keyseven123** [comment]
# Refactor Workflow Build Docker on Demand
As discussed in https://github.com/nebulastream/nebulastream-public/pull/221, we should refactor the workflow for building docker images on demand. IMO, we can take the steps (`docker-development-images, detect-dependency-changes` directly from the `pr.yml` .
